zerver/actions/users.py

```diff
@@ -312,6 +319,102 @@ def do_deactivate_user(
             )
 
 
+def send_stream_events_for_role_update(user_profile: UserProfile, old_role_value: int) -> None:
+    if old_role_value in [
+        UserProfile.ROLE_MEMBER,
+        UserProfile.ROLE_MODERATOR,
+    ] and user_profile.role in [UserProfile.ROLE_MEMBER, UserProfile.ROLE_MODERATOR]:
+        # Stream access permissions are same for member and moderator users.
+        return
+
+    if (
+        old_role_value in [UserProfile.ROLE_REALM_OWNER, UserProfile.ROLE_REALM_ADMINISTRATOR]
+        and user_profile.is_realm_admin
+    ):
+        # Stream access permissions are same for owner and administrator users as well.
+        return
```

Comment:
This is OK, but it's possible it'd be better to just calculate what the user could access before and what they can access now and compare those things -- the logic as written here embeds a lot of understanding of how access permissions work, and with upcoming work on using groups for permissions, those assumptions may no longer be true.

(I guess in particular, probably we'll need to do this sort of diffing whenever changing group memberships).

I'll read the rest of this and we can see if it makes sense to save that for follow-up work though; this adds some nice tests and removes a quirk, so it's OK if there's a potential improvement to the code here that would have the same test results.

Reply:
OK, having read this, it feels like it's a lot of potentially security-sensitive code that I'm not excited about reviewing. Like, I'm very happy to have tests that look at corner cases for particular switches between roles, but I really worry we'll miss some interaction here, because in some sense this function in reimplementing a lot of product access control logic.

I'm more tempted by the idea of doing a `do_get_streams` with the right same parameters used in `/register` (maybe need to extract a helper for that) before changing our `.role`, and again after, and having this logic work by diffing the data calculated using that function?

Reply:
Yes, using some version of `do_get_streams` sounds better. Would try this out and update the PR.